### ISSUE TITLE
MIA: writing output files in "derived_data" folder

### ISSUE THREAD

**Issue by dhrbn** [populse/populse_sandbox#5]
In a MIA project, all the data are stored in a "data" folder which contains three subfolders : 
	- raw_data (for data that come from the import software)
	- downloaded_data (for data that have been imported manually)
	- derived_data (for data that are generated by pipelines)

Most of the processes that have been developed use Nipype's SPM interfaces as below:

    class Smooth(Process_mia):
    
        def __init__(self):
            super(Smooth, self).__init__()
    
            # Inputs
            self.add_trait("in_files", InputMultiPath(ImageFileSPM(), copyfile=False, output=False))
            self.add_trait("fwhm", traits.List([6, 6, 6], output=False, optional=True))
            self.add_trait("data_type", traits.Int(output=False, optional=True))
            self.add_trait("implicit_masking", traits.Bool(output=False, optional=True))
            self.add_trait("out_prefix", traits.String('s', usedefault=True, output=False, optional=True))
    
            # Output
            self.add_trait("smoothed_files", OutputMultiPath(File(), output=True))
    
        def list_outputs(self):
            Process_mia.list_outputs(self)
            process = spm.Smooth()
    
            if not self.in_files:
                return {}
            else:
                process.inputs.in_files = self.in_files
    
            if self.out_prefix:
                process.inputs.out_prefix = self.out_prefix
    
            outputs = process._list_outputs()
    
            inheritance_dict = {}
            for key, values in outputs.items():
                if key == "smoothed_files":
                    for fullname in values:
                        path, filename = os.path.split(fullname)
                        if self.out_prefix:
                            filename_without_prefix = filename[len(self.out_prefix):]
                        else:
                            filename_without_prefix = filename[len('s'):]
    
                        if os.path.join(path, filename_without_prefix) in self.in_files:
                            inheritance_dict[fullname] = os.path.join(path, filename_without_prefix)
    
    
            return outputs, inheritance_dict
    
        def _run_process(self):
            process = spm.Smooth()
            for idx, element in enumerate(self.in_files):
                full_path = os.path.relpath(element)
                self.in_files[idx] = full_path
            process.inputs.in_files = self.in_files
            process.inputs.fwhm = self.fwhm
            process.inputs.data_type = self.data_type
            process.inputs.implicit_masking = self.implicit_masking
            process.inputs.out_prefix = self.out_prefix
    
            process.run()

When such a process is run, the output files are located in the same folder as their input files. And after some research, we have not been able to find a way to change the output directory of a nipype interface. 

We have tried to override the Nipype's Smooth interface and especially its _list_outputs method (that is called when the interface is run) to change the '/raw_data/' of the output files to '/derived_data/'. It returns the desired file names, but it does not affect the behavior of the run (the output files are still in the same folder as the input ones). 

We have also tried to change the output_directory and process_output_directory attributes of StudyConfig but the result is still the same.

What we propose to counter this problem is to copy the input files to the "derived_data" folder just before running a pipeline. It will not affect the database (during the pipeline initialization phase, the "derived_data" file names will be added to the database). 

We know that this is not a "clean" solution beause it will create a lot of new data during the run (the copied files can be removed after that). So if anyone has a better idea to counter the problem we are facing, feel free to propose it!

**Comment by denisri** [comment]
Do all nipype processes behave like this, or just SPM wrappings ? (because it is the way SPM works, but not necessarily other softwxare wrapped in nipype)

On Unix systems it is also possible to make symlinks to original data instead of "heavy" copies. Copying unnecessary data can harm performance in some cases: when using paralllel / distributed processing many processes can run in the same time, then copying files can be an important bottleneck. Moreover if many processes have to copy data this way it is possible to fill a disk and to end up with a disk overflow, so it's worth avoiding when possible. The symlink option is much ighter.

**Comment by servoz** [comment]
@denisri : We only played with SPM wrapping, and it is for this reason that we concluded that it is not allowed to change the output directory (as you pointed out, SPM do not leave possibility to change, from my knowledge). Symbolic link is indeed better to a copy. So we need a a solution for all softwares used in the bricks (as you said witth SPM it seems impossible but with other softwares it is may be possible) and for all platforms (symlinks, for linux i am confident, for mac also but for Windobs I don't know !).
